plugin/acl/setup.go

```diff
@@ -0,0 +1,218 @@
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		a.Next = next
+		return a
+	})
+
+	// Register all metrics.
+	c.OnStartup(func() error {
+		metrics.MustRegister(c, RequestBlockCount, RequestAllowCount)
+		return nil
+	})
+	return nil
+}
+
+func parse(c *caddy.Controller) (acl, error) {
+	a := acl{}
+	for c.Next() {
+		r := Rule{}
+		r.Zones = c.RemainingArgs()
+		if len(r.Zones) == 0 {
+			// if empty, the zones from the configuration block are used.
+			r.Zones = make([]string, len(c.ServerBlockKeys))
+			copy(r.Zones, c.ServerBlockKeys)
+		}
+		for i := range r.Zones {
+			r.Zones[i] = plugin.Host(r.Zones[i]).Normalize()
+		}
+
+		for c.NextBlock() {
+			p := Policy{}
+
+			p.action = strings.ToLower(c.Val())
+			if p.action != ALLOW && p.action != BLOCK {
+				return a, c.Errf("unexpected token '%s'; expect '%s' or '%s'", c.Val(), ALLOW, BLOCK)
+			}
+
+			p.qtypes = make(map[uint16]bool)
+
+			p.filter = iptree.NewTree()
+
+			// match all qtypes and IP addresses.
+			if !c.NextArg() {
+				p.qtypes[QtypeAll] = true
+				p.filter.InplaceInsertNet(IPv4All, true)
+				r.Policies = append(r.Policies, p)
+				break
+			}
+
+			var rawSourceIPRanges []string
+			hasSection := make(map[string]bool)
+			for {
+				section := strings.ToLower(c.Val())
+				tokens, remaining := loadFollowingTokens(c)
+				if len(tokens) == 0 {
+					return a, c.Errf("no token specified in '%s' section", section)
+				}
+
+				switch section {
+				case "type":
+					hasSection["type"] = true
+					for _, token := range tokens {
+						if token == "*" {
+							p.qtypes[QtypeAll] = true
+							continue
+						}
+						qtype, ok := dns.StringToType[token]
+						if !ok {
+							return a, c.Errf("unexpected token '%s'; expect legal QTYPE", token)
+						}
+						p.qtypes[qtype] = true
+					}
+				case "net":
+					hasSection["net"] = true
+					// TODO: refactor to load source IP address one by one.
+					for _, token := range tokens {
+						rawSourceIPRanges = append(rawSourceIPRanges, token)
+					}
+				case "file":
+					hasSection["file"] = true
+					for _, token := range tokens {
+						ipRanges, err := loadNetworksFromLocalFile(token)
+						if err != nil {
+							return a, c.Errf("unable to load networks from local file: %v", err)
+						}
+						for _, ipRange := range ipRanges {
+							rawSourceIPRanges = append(rawSourceIPRanges, ipRange)
+						}
+					}
+				default:
+					return a, c.Errf("unexpected token '%s'; expect 'type | net | file'", c.Val())
+				}
+				if !remaining {
+					break
+				}
+			}
+
+			// optional `type` section means all record types.
+			if !hasSection["type"] {
+				p.qtypes[QtypeAll] = true
+			}
+
+			// optional `net` and `file` means all ip addresses.
+			if !hasSection["net"] && !hasSection["file"] {
+				p.filter.InplaceInsertNet(IPv4All, true)
+			}
+
+			for _, rawNet := range rawSourceIPRanges {
+				if rawNet == "*" {
+					p.filter.InplaceInsertNet(IPv4All, true)
+					continue
```

Comment:
shouldn't break here? No sense to add more networks when we had added ANY network